### ISSUE TITLE
"Crash" when used with prettier and eslint

### ISSUE THREAD

**Issue by SteveALee** [sveltejs/prettier-plugin-svelte#57]
Please see https://github.com/sveltejs/svelte/issues/3550 for details, though is closed there.
I copied the original bug report here, thanks

**Describe the bug**
With [this template](https://github.com/SteveALee/svelte-code-cypress-project) where we want ESLint to co exist with prettier I have 2 problems when the plugins are nebale with [this config](https://github.com/SteveALee/svelte-code-cypress-project/blob/master/.eslintrc.json#L5-L6) 

As described in the [prettier docs ](https://prettier.io/docs/en/integrating-with-linters.html#recommended-configuration) this enable ESLint to call prettier

1) Lint crashes in the svlte compile - see stack trace
2) Prettier fails to format on save the App.svelte file - and error flashes up in the status bar

**Logs**
see stack trace

**To Reproduce**
Uncommented the line aboveand comment the one above it
1) Run npm run lint from the command line
2) edit the App.svelte and save the file

**Expected behavior**
1) no error and llinter succeeds
2) File is reformated via prettier

**Stacktraces**
<details>
  <summary>Stack trace</summary>

$ npm run lint

> svelte-project@1.0.0 lint C:\projects\svelte-code-cypress-project
> eslint src/*

ParseError: Expected }
  1 | export let name
  2 |
> 3 | {name;name=0}
    |     ^
Occurred while linting C:\projects\svelte-code-cypress-project\src\App.svelte:3
    at error$1 (C:\projects\svelte-code-cypress-project\node_modules\svelte\compiler.js:13329:20)
    at Parser$2.error (C:\projects\svelte-code-cypress-project\node_modules\svelte\compiler.js:13405:10)
    at Parser$2.eat (C:\projects\svelte-code-cypress-project\node_modules\svelte\compiler.js:13419:19)
    at mustache (C:\projects\svelte-code-cypress-project\node_modules\svelte\compiler.js:13220:17)
    at new Parser$2 (C:\projects\svelte-code-cypress-project\node_modules\svelte\compiler.js:13364:22)
    at Object.parse$1 [as parse] (C:\projects\svelte-code-cypress-project\node_modules\svelte\compiler.js:13495:21)
    at Object.parse (C:\projects\svelte-code-cypress-project\node_modules\prettier-plugin-svelte\plugin.js:669:51)
    at Object.parse$2 [as parse] (C:\projects\svelte-code-cypress-project\node_modules\prettier\index.js:10629:19)
    at coreFormat (C:\projects\svelte-code-cypress-project\node_modules\prettier\index.js:13888:23)
    at format (C:\projects\svelte-code-cypress-project\node_modules\prettier\index.js:14146:73)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! svelte-project@1.0.0 lint: `eslint src/*`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the svelte-project@1.0.0 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\steve\AppData\Roaming\npm-cache\_logs\2019-09-11T17_00_42_583Z-debug.log
</details>

**Information about your Svelte project:**
- Your browser and the version: (e.x. Chrome 52.1, Firefox 48.0, IE 10)
Not relevant - build error

- Your operating system: (e.x. OS X 10, Ubuntu Linux 19.10, Windows XP, etc)
Windows 10 latest 1903

- Svelte version (Please check you can reproduce the issue with the latest release!)
3.12.1

- Whether your project uses Webpack or Rollup
Roll up

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Svelte entirely?

Serious - can't lint or run prettier on svelte files

**Additional context**

**Comment by btamas** [comment]
I faced with the same issue. Any update on this?

**Comment by jzillmann** [comment]
Me too!

**Comment by toerndev** [comment]
It seems like prettier-plugin-svelte isn't compatible with eslint-plugin-svelte3, however meanwhile we can still extend eslint-config-prettier and _separately_ call prettier.
For those using vim and w0rp/ale, this works with `let g_ale_linter_aliases = { 'svelte': ['javascript'] }` and `let g:ale_fixers = { ..., 'svelte': ['eslint', 'prettier'] }` (eslint as both linter and fixer to get both warnings and fixes). Also with w0rp/ale it replaces the buffer with any log output from prettier on syntax error (!), so open `node_modules/prettier-plugin-svelte/plugin.js` and comment out the `console.log` line.

**Comment by maxwell-k** [comment]
I have the same problem, which means I am not using `eslint-plugin-prettier` although I would like to. I am also using vim and ALE but have a slightly different workaround; I don't get fixes from eslint,  but for me this is perhaps simpler:

```vim
let b:ale_linter_aliases = {'svelte': 'javascript'}
let b:ale_linters = ['eslint',]
let b:ale_fixers = ['prettier',]
```

I haven't got as far as figuring out where to start with a fix for the underlying issue.

**Comment by toerndev** [comment]
Did some debugging, seems like `eslint-plugin-svelte3` only extracts JS from the `<script>` part and returns this (through `eslint-plugin-prettier`) for prettier to parse, meanwhile `prettier-plugin-svelte` expects a whole .svelte file to break down into an AST, and crashes.

I'm trying to get this confirmed in the Prettier community but it seems to me like the eslint plugin is at fault here for not sending the whole file, which Prettier *is* capable of parsing (with the plugin). The JSX in the template part of .svelte files should also be formatted by Prettier, and if we look at `eslint-plugin-react` it indeed sends the whole .jsx file through `eslint-plugin-prettier`.

I'd like to ask the Svelte community why they say that `prettier-plugin-svelte` is wrong (see issue #3550 linked above) + that we should disable `eslint-plugin-prettier` when using their eslint plugin. Is it a misunderstanding, or was it just too difficult to implement the eslint plugin differently at the time...?

**Comment by SteveALee** [comment]
@torndev I found the same and worked around this in VS Code - 

see https://github.com/SteveALee/svelte-code-cypress-project/blob/master/README.md

I created a bug with more details here https://github.com/sveltejs/svelte/issues/3550 but seems it was the wrong place.

Also https://github.com/UnwrittenFun/svelte-vscode/issues/74

Also see

**Comment by toerndev** [comment]
@SteveALee Yeah the answer to [sveltejs/svelte#3550](https://github.com/sveltejs/svelte/issues/3550) is what's making me confused. It would seem like exactly the right place to resolve the problem. Maybe we can try to get it reopened.

**Comment by toerndev** [comment]
Thought I'd share my findings so far. Perhaps someone knows more and/or would like to join the search for a solution! I haven't worked with eslint plugins before, I'm just digging in node_modules trying to trace the plugin call order.

We have eslint calling `eslint-plugin-svelte3` which has `preprocess, postprocess` functions (see _Working with Plugins_ in ESLint's developer guide). `preprocess` calls `svelte3/compiler` to generate an AST, and returns an extract of the pure JS parts.
The strings returned by `eslint-plugin-svelte3: preprocess` are then forwarded `eslint-plugin-prettier -> prettier -> prettier-plugin-svelte` which was expecting a .svelte file and crashes.

The maintainers of `eslint-plugin-svelte3` (which is the official plugin maintained by the Svelte team) have pointed at this plugin and didn't seem very interested in discussing another approach. But I believe it's the eslint plugin that's not behaving as expected in the big picture.
So although I haven't ever used Vue yet, I looked into the Vue plugin stack for eslint/prettier, because its file structure looks very similar to Svelte's: JSX and script + style + template.
What Vue does, is that eslint calls `vue-eslint-parser` (see ESLint's _Working with Custom Parsers_), then `eslint-plugin-vue: preprocess` which just returns the entire .vue file's contents which is then sent  to their Prettier plugin.

Next step is to determine how hard it would be to move the `svelte3/compiler` AST stuff to a separate _parser_ library, and have the _plugin_ be much simpler, just like `eslint-plugin-vue` which mostly contains lots of rule definitions and only 70 lines for the actual processors.
The Vue parser & plugin seem to be written by eslint maintainers though, so in the worst case this approach requires official support from eslint.
Also I could only see `vue-eslint-parser` being called for files _imported by_ the main .vue file, so it's still unclear where the main file's AST is built.

TL;DR: lots of intrigue simply because I don't know how any of this works... :laughing:

**Comment by boojum** [comment]
I think that I'm facing the same issue. I'm linting with `eslint` taking advantage of the `prettier/recommended` rules set. After adding `eslint-plugin-svelte3` everything was still working as it should. As soon as I added `prettier-plugin-svelte` it stopped. This snippet:

```
<script>
  export let name
</script>
```
... is causing `eslint` to fail with:

```
npx eslint --fix ./src/**/*.{js,svelte}
Expected }
  1 | export let name
  2 | 
> 3 | {name;name=0}
    |     ^
Occurred while linting /home/boojum/code/playground/parcel-svelte/src/containers/App/App.svelte:3
```

Repo with the minimal setup required to reproduce is [here](https://gitlab.com/b00jum/broken-prettier-recommend). To reproduce, simply run `npx eslint --fix ./src/**/*.{js,svelte}` - it will fail with the above error. Then uninstal `prettier-plugin-svelte` and run `eslint` again - no problems at all.

**Comment by dummdidumm** [comment]
As stated in the [newest Prettier docs](https://prettier.io/docs/en/integrating-with-linters.html), using packages like `eslint-plugin-prettier` is no longer recommended. Instead, use a preset configuration which turns off all Eslint rules which conflict with prettier and then just format first with prettier and run the linter afterwards.

**Comment by dionysiusmarquis** [comment]
> As stated in the [newest Prettier docs](https://prettier.io/docs/en/integrating-with-linters.html), using packages like `eslint-plugin-prettier` is no longer recommended…

@dummdidumm But only eslint plugins provide "realtime" linting inside the IDE, or not? In my workflow I'm gettung linting errors etc. highlighted while i'm typing. With the plugin I also get prettier "errors" highlighted while typing. It's a thing of preference, but I prefer it this way instead of constantly getting my code reformatted on save e.g.

**Comment by Nick-Mazuk** [comment]
I'm getting this error, too. It would be really annoying to use Prettier separately because every single non-svelte project has Prettier integrated with ESLint. So Svelte projects are the only projects requiring me to setup extra tooling.

https://stackoverflow.com/questions/66804739/using-prettier-plugins-with-eslint-plugin-prettier?noredirect=1#comment118107753_66804739

**Comment by dummdidumm** [comment]
The problem is neither `eslint-plugin-svelte3` nor `prettier-plugin-svelte`, it's `prettier-plugin-svelte` not making them play nice together. The problem is that `eslint-plugin-svelte3` needs to extract the JS parts of the Svelte file, because ESLint can only run on that. This means that something like `export let foo` is passed to `prettier-plugin-svelte`, which then throws, because that is invalid Svelte code. Looking at the code [it seems that enhancing this line](https://github.com/prettier/eslint-plugin-prettier/blob/master/eslint-plugin-prettier.js#L168) with `.svelte` may be enough (maybe that should also be made configurable on their side). I suggest you try that out and see if my hunch is correct. I'll also open an issue over there.

**Comment by dclasen-tarent** [comment]
@dummdidumm I tried your suggestion and it seems to resolve the crash issue. However the linter still reports false positive errors for prettier/prettier:

```
<script lang="ts">
    interface Foo {
      baz: string[];
      foo: boolean;
    }

    export let properties: Foo
</script>

{#if properties.foo}
    {#each properties.baz as bar}
      <p>
        {bar}
      </p>
    {/each}
{/if}
```

```
Foo.svelte
  11:12  error  Insert `··`  prettier/prettier
```

That is the line: `{#each properties.baz as bar}` where eslint/prettier wants to fix it with extra whitepaces between the `#each` and `properties.baz` which does not help to get rid of the error. It will just keep complaining and --fix will just keep adding more whitespaces as prettier is still not running on the source but the transpiled code given from the `eslint-plugin-svelte3`.

**Comment by JounQin** [comment]
From https://github.com/prettier/eslint-plugin-prettier/issues/472

> It should be easy to fix this by adding `svelte` into
>
> https://github.com/prettier/eslint-plugin-prettier/blob/551dd0e29840f189df71bfc9bd3fd5aa74f2b8f1/eslint-plugin-prettier.js#L215-L225
>
> Would you like to raise a PR?

**Comment by JounQin** [comment]
I think this can be closed in favor of https://github.com/prettier/eslint-plugin-prettier/pull/483

**Comment by piyushere** [comment]
> I think this can be closed in favor of https://github.com/prettier/eslint-plugin-prettier/pull/483

I don't think so, I am still getting this error with the _eslint-plugin-prettier@4.2.1_ and the official _eslint-plugin-svelte3_:
![image](https://user-images.githubusercontent.com/23546470/201672850-1f332ab7-9f99-431e-814c-6e0a8fff5719.png)
which doesn't make any sense.

I have not tried this with the recommended _eslint-plugin-svelte_ package yet.
